### PR TITLE
WIP: OpenUrl handles FileDrop for file opening.

### PR DESCRIPTION
**Pull request by MiraiSubject**
### Purpose of this PR

* Calls the filedrop eventhandler for osu.game to process

### Testing status

* Untested
* Builds and runs SampleGame.iOS successfully with the assemblies built from this PR.
* Builds and runs osu.iOS successfully with the osu.iOS assembly
### Comments

Needs a change of inheritance for `GameAppDelegate.cs` in osu.Framework.iOS.
[ppy/osu-framework#2048](https://github.com/ppy/osu-framework/pull/2048) is adapted for this change.

Basically 

`public abstract class GameAppDelegate : UIApplicationDelegate`
becomes: 
`public abstract class GameAppDelegate : IOSAppDelegate`

This invalidates [ppy/osu#3914](https://github.com/ppy/osu/pull/3914).

Looking at the game code a separate PR might be required to properly handle the event in-game, because the event is handled in osu.Desktop's `OsuGameDesktop.cs`.

Would like to get suggestions on how to get the implementation in the actual game without touching the osu.iOS project.

`src/osuTK/Input/FileDropEventArgs.cs` had to change from `internal set;` to `set;`, because else osuTK.iOS would complain about it being inaccessible because it is a different assembly.

### CODE REVIEW

**Comment by swoolcock** [comment]
@MiraiSubject I'll give you a hand with this later if you want, just ping me on Discord. (Samah)

**Comment by peppy** [comment]
Unsure this is the correct direction. File drops are different from URL handlers. Even on windows.

**Comment by swoolcock** [comment]
I was hoping to keep it at an osuTK level rather than osu-framework, and osuTK does not have the concept of URL handlers (though it could be added).

It's fine to add it to osu-framework instead, I just thought it would be nice to make it at a lower level.

**Comment by MiraiSubject** [comment]
I don't know which direction we should be working in, but it is correct that iOS's URL handler is different from a `FileDrop`.

On iOS files go through a URL handler in the form of `file://(path from root to app container)`. If a file is sent to the app it gets copied from its original location (like the safari download cache) to  `<app-container>/Documents/Inbox/`. 

In that same method the actual URL handler for stuff like osu!direct `osu://` could be checked too using for example regex because it's all stored in in the `url` variable from the method. 

So the thought process was to fire the `FileDrop` event in the case a file was sent to the app in the URL handler and if something from `osu://` was sent we would be able to handle that differently. 

Something that should also be kept in mind (which wasn't implemented yet) were cases of failed imports, they stay in `<app-container>/Documents/Inbox/` and has to be cleaned manually.

I hope this clears up the whole URL handler situation, because that's just a quirk specific to iOS (at least I don't know if any other platform handles it like this)

File: src/osuTK.iOS/IOSAppDelegate.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+using Foundation;
+using osuTK.Input;
+using osuTK.Platform;
+
+namespace osuTK.iOS
+{
+    public abstract class IOSAppDelegate : UIApplicationDelegate
+    {
+       
+        public abstract override bool FinishedLaunching(UIApplication application, NSDictionary launchOptions); // Exposing FinishedLaunching for osu-framework to handle
```
Comment (peppy) [inline src/osuTK.iOS/IOSAppDelegate.cs]: this is never being called?

File: src/osuTK.iOS/IOSAppDelegate.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+        private readonly FileDropEventArgs FileDropArgs = new FileDropEventArgs();
+
+
+
+        public override bool OpenUrl(UIApplication app, NSUrl url, NSDictionary options) // File opening from external source. (Required Info.plist in game project)
+        {
+            var e = FileDropArgs;
```
Comment (peppy) [inline src/osuTK.iOS/IOSAppDelegate.cs]: this is just wrong. why aren't you making a new instance of `FileDropEventArgs`?

File: src/osuTK/Input/FileDropEventArgs.cs [hunk 3]
```diff
@@ -14,7 +14,7 @@ public class FileDropEventArgs : EventArgs
         public string FileName
         {
             get;
-            internal set;
+            set;
```
Comment (peppy) [inline src/osuTK/Input/FileDropEventArgs.cs:17]: why is this required?

File: src/osuTK.iOS/IOSAppDelegate.cs (hunk 1 above)
Comment (MiraiSubject) [reply src/osuTK.iOS/IOSAppDelegate.cs]: It is in osu-framework, that’s a change that is needed for the game to boot at all [osu-framework#2048](
https://github.com/ppy/osu-framework/pull/2048/commits/7e333c759156a47cedb7a2b4460e321fe7f1db88)

Comment (peppy) [reply src/osuTK.iOS/IOSAppDelegate.cs]: Then include it in osu-framework, not here...

File: src/osuTK.iOS/IOSAppDelegate.cs (hunk 2 above)
Comment (MiraiSubject) [reply src/osuTK.iOS/IOSAppDelegate.cs]: That was indeed incorrect. When I made this commit did not know how the eventhandler worked.

File: src/osuTK/Input/FileDropEventArgs.cs (hunk 3 above)
Comment (MiraiSubject) [reply src/osuTK/Input/FileDropEventArgs.cs:17]: This was still a part of my incorrect understanding of the eventhandler. I have reverted this change.

**Comment by MiraiSubject** [comment]
The current implementation now makes the FileDrop event working on iOS. This event gets fired in [ppy/osu-framework#2048 -> GameAppDelegate.cs](https://github.com/ppy/osu-framework/blob/728b2c84150067fdfe40b6d4150a4a9838a71722/osu.Framework.iOS/GameAppDelegate.cs) for handling files through iOS's URL and Filehandler `OpenUrl`

Edit: The firing of the event is testable in SampleGame.iOS

Edit 2: Seems I do require "internal set" -> "set" 
`error CS0200: Property or indexer 'FileDropEventArgs.FileName' cannot be assigned to -- it is read only [C:\projects\opentk\src\osuTK.iOS\osuTK.iOS.csproj]` Good of Appveyor to catch that cause somehow building it in MSVC didn't throw this (probably still built it from the cache despite the change)

It is needed due to osuTK.NS20 being in a different assembly from osuTK.iOS, so that kind of makes sense.

File: src/osuTK.iOS/iOSGameView.cs [hunk 4]
```diff
@@ -975,7 +988,8 @@ protected virtual void OnUnload(EventArgs e)
         public event EventHandler<MouseButtonEventArgs> MouseUp;
         public event EventHandler<MouseMoveEventArgs> MouseMove;
         public event EventHandler<MouseWheelEventArgs> MouseWheel;
-        public event EventHandler<FileDropEventArgs> FileDrop;
+        public event EventHandler<FileDropEventArgs> FileDrop = delegate { };
```
Comment (swoolcock) [inline src/osuTK.iOS/iOSGameView.cs]: Why?

Comment (MiraiSubject) [reply src/osuTK.iOS/iOSGameView.cs]: I assumed from `NativeWindow.cs` the same approach would've been required. Upon further testing without this. It seems like this change was not needed. A push will follow reverting this change.

File: src/osuTK.iOS/iOSGameView.cs [hunk 5]
```diff
@@ -952,6 +962,9 @@ protected virtual void OnUnload(EventArgs e)
 
         protected virtual void OnRenderFrame(FrameEventArgs e) => RenderFrame?.Invoke(this, e);
 
+
```
Comment (peppy) [inline src/osuTK.iOS/iOSGameView.cs]: unnecessary newlines

File: src/osuTK.iOS/iOSGameView.cs [hunk 6]
```diff
@@ -977,6 +990,7 @@ protected virtual void OnUnload(EventArgs e)
         public event EventHandler<MouseWheelEventArgs> MouseWheel;
         public event EventHandler<FileDropEventArgs> FileDrop;
 
+
```
Comment (peppy) [inline src/osuTK.iOS/iOSGameView.cs]: unnecessary newlines

File: src/osuTK.iOS/iOSGameView.cs [hunk 7]
```diff
@@ -712,6 +712,16 @@ public override void LayoutSubviews()
             }
         }
 
+
+        private readonly FileDropEventArgs FileDropArgs = new FileDropEventArgs();
```
Comment (peppy) [inline src/osuTK.iOS/iOSGameView.cs]: again, this is wrong. make it where you use it.

File: src/osuTK.iOS/iOSGameView.cs [hunk 8]
```diff
@@ -712,6 +712,16 @@ public override void LayoutSubviews()
...
+        public void OnFileDrop(string s)
+        {
+            Console.WriteLine("Importing " + s);
```
Comment (peppy) [inline src/osuTK.iOS/iOSGameView.cs]: please don't writeline to console in libraries.

**Review by peppy — CHANGES_REQUESTED** [review]

File: src/osuTK.iOS/iOSGameView.cs (hunk 7 above)
Comment (MiraiSubject) [reply src/osuTK.iOS/iOSGameView.cs]: I make it when I use it though, if I understand what you mean: 

`
private readonly FileDropEventArgs FileDropArgs = new FileDropEventArgs();
        public void OnFileDrop(string s)
        {
            var e = FileDropArgs; //FileDropArgs used here
            FileDropArgs.FileName = s;
            FileDrop(this, e);
        }
`

This is also the same case for `osuTK.NS20/Platform/NativeWindowBase.cs`
https://github.com/ppy/osuTK/blob/055a74f36cff552f7200093db032a3a9f4351591/src/osuTK/Platform/NativeWindowBase.cs#L48
https://github.com/ppy/osuTK/blob/055a74f36cff552f7200093db032a3a9f4351591/src/osuTK/Platform/NativeWindowBase.cs#L148
https://github.com/ppy/osuTK/blob/055a74f36cff552f7200093db032a3a9f4351591/src/osuTK/Platform/NativeWindowBase.cs#L150
  where it's also declared within the same class and used like that.

**Comment by MiraiSubject** [comment]
Replaced 

`private readonly FileDropEventArgs FileDropArgs = new FileDropEventArgs(); `
with: 

`var e = new FileDropEventArgs();`

in the body instead of globally in the class, so a new instance is created on every call.

Also did the other changes as requested.

File: src/osuTK/Input/FileDropEventArgs.cs (hunk 3 above)
Comment (peppy) [reply src/osuTK/Input/FileDropEventArgs.cs:17]: doesn't look to be reverted.

File: src/osuTK.iOS/iOSGameView.cs [hunk 9]
```diff
@@ -712,6 +712,14 @@ public override void LayoutSubviews()
             }
         }
 
+
+        public void OnFileDrop(string s)
```
Comment (peppy) [inline src/osuTK.iOS/iOSGameView.cs:716]: this is never being called..

File: src/osuTK.iOS/iOSGameView.cs [hunk 10]
```diff
@@ -952,6 +960,7 @@ protected virtual void OnUnload(EventArgs e)
 
         protected virtual void OnRenderFrame(FrameEventArgs e) => RenderFrame?.Invoke(this, e);
 
+
```
Comment (peppy) [inline src/osuTK.iOS/iOSGameView.cs:963]: newline

File: src/osuTK/Input/FileDropEventArgs.cs (hunk 3 above)
Comment (MiraiSubject) [reply src/osuTK/Input/FileDropEventArgs.cs:17]: See the edit in my comment below as to why it wasn't reverted: 

> Edit 2: Seems I do require "internal set" -> "set"
> `error CS0200: Property or indexer 'FileDropEventArgs.FileName' cannot be assigned to -- it is read only [C:\projects\opentk\src\osuTK.iOS\osuTK.iOS.csproj]` Good of Appveyor to catch that cause somehow building it in MSVC didn't throw this (probably still built it from the cache despite the change)
> 
> It is needed due to osuTK.NS20 being in a different assembly from osuTK.iOS, so that kind of makes sense.

FileDropEventArgs also inherits from EventArgs so if you want to keep it `internal` that + FileDropEventArgs have to be recreated in the osuTK.iOS project

File: src/osuTK.iOS/iOSGameView.cs (hunk 9 above)
Comment (MiraiSubject) [reply src/osuTK.iOS/iOSGameView.cs:716]: it's getting called here for the first time to handle FileDrops: [ppy/osu-framework#2048 -> GameAppDelegate.cs](https://github.com/ppy/osu-framework/blob/d92ee47e1207951b2490d63e2a0ead1123914f1c/osu.Framework.iOS/GameAppDelegate.cs#L64)

Comment (peppy) [reply src/osuTK.iOS/iOSGameView.cs:716]: This is just too convoluted.

Comment (MiraiSubject) [reply src/osuTK.iOS/iOSGameView.cs:716]: It's possible to call it in osuTK, but the AppDelegate inherited from UIApplicationDelegate will then have to move over here and that might conflict with whats going on inside the framework.

Comment (peppy) [reply src/osuTK.iOS/iOSGameView.cs:716]: put it this way: doing it as you have it now is objectively wrong. it either needs to be called here or it needs to note exist here.
